pyparsing/diagram/__init__.py

```diff
@@ -200,17 +214,30 @@ class ConverterState:
     Stores some state that persists between recursions into the element tree
     """
 
-    def __init__(self, diagram_kwargs: dict = {}):
+    def __init__(self, diagram_kwargs: Optional[Dict] = None):
         #: A dictionary mapping ParserElement IDs to state relating to them
-        self.first = {}  # type:  Dict[int, ElementState]
+        self._first: Dict[int, ElementState] = {}
         #: A dictionary mapping ParserElement IDs to subdiagrams generated from them
-        self.diagrams = {}  # type: Dict[int, EditablePartial[NamedDiagram]]
+        self.diagrams: Dict[int, EditablePartial[NamedDiagram]] = {}
         #: The index of the next unnamed element
-        self.unnamed_index = 1  # type:  int
+        self.unnamed_index: int = 1
         #: The index of the next element. This is used for sorting
-        self.index = 0  # type:  int
+        self.index: int = 0
         #: Shared kwargs that are used to customize the construction of diagrams
-        self.diagram_kwargs = diagram_kwargs  # type:  dict
+        self.diagram_kwargs: Dict = diagram_kwargs or {}
+        self.extracted_diagram_names: Set[str] = set()
+
+    def __setitem__(self, key, value):
```

Comment:
This is a fair change, since we mostly interact with this class's `first` attribute. Maybe add types for the params and return values here?

Reply:
The name "first" just generally confused me, so getitem/setitem kept the logic and made the calling code more natural. Adding type annotations in next update.

Reply:
In that case I'm happy for you to change the name of the `first` attribute as well. I can't actually remember why I called it that, possibly because it stores the "first" copy of each element we encounter. But then it sounds like elements are copied when reused so this doesn't really make sense anyway.